File: backend/html/src/main/java/com/vladsch/flexmark/util/sequence/HackReplacer.java
package com.vladsch.flexmark.util.sequence;

public final class HackReplacer implements Escaping.Replacer {
    public void replace(String s, StringBuilder sb) {
        switch (s) {
            case "&" -> {
                sb.append("&amp;");
            }
            case "<" -> {
                sb.append("&lt;");
            }
            case ">" -> {
                sb.append("&gt;");
            }
            default -> {
                sb.append(s);
            }
        }
    }

    public void replace(BasedSequence original, int startIndex, int endIndex, ReplacedTextMapper textMapper) {

        String value = original.subSequence(startIndex, endIndex).toString();
        switch (value) {
            case "&" -> {
                textMapper.addReplacedText(startIndex, endIndex, PrefixedSubSequence.prefixOf("&amp;", BasedSequence.NULL));
            }
            case "<" -> {
                textMapper.addReplacedText(startIndex, endIndex, PrefixedSubSequence.prefixOf("&lt;", BasedSequence.NULL));
            }
            case ">" -> {
                textMapper.addReplacedText(startIndex, endIndex, PrefixedSubSequence.prefixOf("&gt;", BasedSequence.NULL));
            }
            default -> {
                textMapper.addOriginalText(startIndex, endIndex);
            }
        }
    }
}

File: backend/src/main/kotlin/com/vladsch/flexmark/util/sequence/HackReplacer.kt
package com.vladsch.flexmark.util.sequence

class HackReplacer : Escaping.Replacer {
    override fun replace(s: String, sb: StringBuilder) {
        when (s) {
            "&" -> {
                sb.append("&amp;")
            }
            "<" -> {
                sb.append("&lt;")
            }
            ">" -> {
                sb.append("&gt;")
            }
            else -> {
                sb.append(s)
            }
        }
    }

    override fun replace(original: BasedSequence, startIndex: Int, endIndex: Int, textMapper: ReplacedTextMapper) {
        when (original.subSequence(startIndex, endIndex).toString()) {
            "&" -> {
                textMapper.addReplacedText(startIndex, endIndex, PrefixedSubSequence.prefixOf("&amp;", BasedSequence.NULL))
            }
            "<" -> {
                textMapper.addReplacedText(startIndex, endIndex, PrefixedSubSequence.prefixOf("&lt;", BasedSequence.NULL))
            }
            ">" -> {
                textMapper.addReplacedText(startIndex, endIndex, PrefixedSubSequence.prefixOf("&gt;", BasedSequence.NULL))
            }
            else -> {
                textMapper.addOriginalText(startIndex, endIndex)
            }
        }
    }
}